File: scripts/daily-review/backfill-limit-up.py
"""limitUpReasons 历史回填脚本

用法：python3 backfill-limit-up.py [--from 2026-02-13] [--to 2026-04-13] [--dry-run]

逐日调用 jiuyan-image-fetch.ts 拉取涨停简图并写入 DB。
幂等：已存在的日期自动跳过。周末自动跳过。
"""

import argparse
import json
import subprocess
import sys
import time
import uuid
from datetime import datetime, timedelta, timezone
from pathlib import Path

# 加载环境变量
from dotenv import load_dotenv
env_path = Path(__file__).resolve().parent.parent.parent / 'apps' / 'web' / '.env.local'
load_dotenv(env_path)

import os
from supabase import create_client

url = os.environ.get('SUPABASE_URL') or os.environ.get('NEXT_PUBLIC_SUPABASE_URL', '')
key = os.environ.get('SUPABASE_ANON_KEY') or os.environ.get('NEXT_PUBLIC_SUPABASE_ANON_KEY', '')
if not url or not key:
    print('[error] 缺少 Supabase 环境变量')
    sys.exit(1)

sb = create_client(url, key)

SCRIPT_DIR = Path(__file__).resolve().parent
FETCH_SCRIPT = SCRIPT_DIR / 'jiuyan-image-fetch.ts'

# 2026 年 A 股休市日（春节、清明、劳动节等）
HOLIDAYS_2026 = {
    # 春节 2/14-2/20（调休 2/14 周六、2/15 周日上班但仍休市到 2/22）
    '2026-02-14', '2026-02-15', '2026-02-16', '2026-02-17',
    '2026-02-18', '2026-02-19', '2026-02-20',
    # 清明 4/4-4/6
    '2026-04-04', '2026-04-05', '2026-04-06',
}


def is_non_trading(date_str: str) -> bool:
    d = datetime.strptime(date_str, '%Y-%m-%d')
    return d.weekday() >= 5 or date_str in HOLIDAYS_2026


def generate_dates(from_date: str, to_date: str) -> list[str]:
    dates = []
    cur = datetime.strptime(from_date, '%Y-%m-%d')
    end = datetime.strptime(to_date, '%Y-%m-%d')
    while cur <= end:
        ds = cur.strftime('%Y-%m-%d')
        if not is_non_trading(ds):
            dates.append(ds)
        cur += timedelta(days=1)
    return dates


def main():
    parser = argparse.ArgumentParser(description='limitUpReasons 历史回填')
    parser.add_argument('--from', dest='from_date', default='2026-02-13')
    parser.add_argument('--to', dest='to_date', default='2026-04-13')
    parser.add_argument('--dry-run', action='store_true')
    args = parser.parse_args()

    print(f'limitUpReasons 历史回填')
    print(f'范围: {args.from_date} ~ {args.to_date}')
    if args.dry_run:
        print('(dry-run 模式)')
    print()

    # 查已有日期
    existing_dates = set()
    offset = 0
    while True:
        resp = sb.table('limitUpReasons').select('pick_date').range(offset, offset + 999).execute()
        if not resp.data:
            break
        for r in resp.data:
            existing_dates.add(r['pick_date'])
        if len(resp.data) < 1000:
            break
        offset += 1000

    print(f'DB 已有 {len(existing_dates)} 天数据')

    # 生成日期列表
    all_dates = generate_dates(args.from_date, args.to_date)
    dates = [d for d in all_dates if d not in existing_dates]

    print(f'需回填: {len(dates)} 天'
          f'（跳过 {len(all_dates) - len(dates)} 天：周末+已有）\n')

    if not dates:
        print('无需回填')
        return

    if args.dry_run:
        for i, d in enumerate(dates):
            print(f'  [{i+1}/{len(dates)}] {d} (dry-run)')
        return

    success = 0
    fail = 0
    skip = 0
    consecutive_fails = 0

    for i, date in enumerate(dates):
        print(f'[{i+1}/{len(dates)}] {date}...', end=' ', flush=True)

        try:
            proc = subprocess.run(
                ['npx', '--yes', 'tsx', str(FETCH_SCRIPT), date],
                cwd=str(SCRIPT_DIR),
                capture_output=True,
                text=True,
                timeout=300,
                env={**os.environ},
            )

            if proc.returncode != 0:
                stderr = proc.stderr.strip()
                # API 返回 errCode=0 但无图片 → 非交易日，不算真正失败
                if 'errCode=0' in stderr:
                    print('无涨停简图（非交易日/假期），跳过')
                    skip += 1
                    consecutive_fails = 0
                    time.sleep(1)
                    continue
                print(f'脚本失败: {stderr[:150]}')
                fail += 1
                consecutive_fails += 1
                if consecutive_fails >= 3:
                    print('\n连续失败 3 次，停止。请检查 JIUYAN_SESSION。')
                    break
                time.sleep(5)
                continue

            data = json.loads(proc.stdout.strip())
            themes = data.get('themes', [])

            if not themes:
                print('无题材（非交易日），跳过')
                skip += 1
                consecutive_fails = 0
                time.sleep(1)
                continue

            # 写入 DB
            record = {
                'id': str(uuid.uuid4()),
                'pick_date': date,
                'themes': themes,
                'raw_image_url': data.get('raw_image_url'),
                'source': 'jiuyan-image-backfill',
                'created_at': int(time.time() * 1000),
            }
            sb.table('limitUpReasons').insert(record).execute()
            print(f'成功: {data["theme_count"]} 题材, {data["stock_count"]} 股票')
            success += 1
            consecutive_fails = 0

            # 间隔 3 秒
            if i < len(dates) - 1:
                time.sleep(3)

        except subprocess.TimeoutExpired:
            print('超时(300s)')
            fail += 1
            consecutive_fails += 1
            if consecutive_fails >= 3:
                print('\n连续失败 3 次，停止。')
                break
            time.sleep(5)
        except json.JSONDecodeError as e:
            print(f'JSON 解析失败: {e}')
            fail += 1
            consecutive_fails += 1
            time.sleep(5)
        except Exception as e:
            print(f'异常: {str(e)[:100]}')
            fail += 1
            consecutive_fails += 1
            time.sleep(5)

    print(f'\n回填完成: 成功 {success}, 失败 {fail}, 跳过(非交易日) {skip}')


if __name__ == '__main__':
    main()

File: scripts/daily-review/backfill-limit-up.ts
/**
 * limitUpReasons 历史回填脚本
 *
 * 用法：npx tsx backfill-limit-up.ts [--from 2026-02-13] [--to 2026-04-13] [--dry-run]
 *
 * 逐日调用 jiuyan-image-fetch.ts 拉取涨停简图并写入 DB。
 * 幂等：已存在的日期自动跳过。
 * 周末自动跳过（周六日非交易日）。
 *
 * 环境变量：
 *   JIUYAN_SESSION / ANTHROPIC_AUTH_TOKEN / ANTHROPIC_BASE_URL
 *   SUPABASE_URL / NEXT_PUBLIC_SUPABASE_URL
 *   SUPABASE_SERVICE_KEY（优先）/ SUPABASE_ANON_KEY
 */

import { resolve } from 'path';
import { spawnSync } from 'child_process';
import { config } from 'dotenv';
import { createClient } from '@supabase/supabase-js';
import { randomUUID } from 'crypto';

// 加载环境变量
const scriptDir = import.meta.dirname ?? process.cwd();
const envPath = resolve(scriptDir, '../../apps/web/.env.local');
config({ path: envPath });

const supabaseUrl =
  process.env.SUPABASE_URL || process.env.NEXT_PUBLIC_SUPABASE_URL || '';
const supabaseKey =
  process.env.SUPABASE_SERVICE_KEY || process.env.SUPABASE_ANON_KEY || process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY || '';

if (!supabaseUrl || !supabaseKey) {
  console.error('[error] 缺少 Supabase 环境变量');
  process.exit(1);
}

const sb = createClient(supabaseUrl, supabaseKey);

// 解析命令行参数
const args = process.argv.slice(2);
function getArg(name: string, defaultVal: string): string {
  const idx = args.indexOf(name);
  return idx !== -1 && args[idx + 1] ? args[idx + 1] : defaultVal;
}
const dryRun = args.includes('--dry-run');

// 默认：从最早数据倒推60个交易日 ≈ 85自然日
const defaultFrom = '2026-02-13';
const defaultTo = '2026-04-13';
const fromDate = getArg('--from', defaultFrom);
const toDate = getArg('--to', defaultTo);

const DATE_PATTERN = /^\d{4}-\d{2}-\d{2}$/;
if (!DATE_PATTERN.test(fromDate) || !DATE_PATTERN.test(toDate)) {
  console.error('[error] 日期参数格式必须为 YYYY-MM-DD');
  process.exit(1);
}

function addDays(dateStr: string, days: number): string {
  const d = new Date(dateStr + 'T00:00:00+08:00');
  d.setDate(d.getDate() + days);
  return d.toISOString().slice(0, 10);
}

function isWeekend(dateStr: string): boolean {
  const d = new Date(dateStr + 'T12:00:00+08:00');
  return d.getDay() === 0 || d.getDay() === 6;
}

async function main() {
  console.log(`limitUpReasons 历史回填`);
  console.log(`范围: ${fromDate} ~ ${toDate}`);
  if (dryRun) console.log('(dry-run 模式，不写入)');
  console.log('');

  // 查已有日期
  const { data: existing } = await sb
    .from('limitUpReasons')
    .select('pick_date');
  const existingDates = new Set((existing || []).map(r => r.pick_date));
  console.log(`DB 已有 ${existingDates.size} 天数据`);

  // 生成日期列表
  const dates: string[] = [];
  let cur = fromDate;
  while (cur <= toDate) {
    if (!isWeekend(cur) && !existingDates.has(cur)) {
      dates.push(cur);
    }
    cur = addDays(cur, 1);
  }

  console.log(`需回填: ${dates.length} 天（已跳过周末和已有日期）\n`);

  if (dates.length === 0) {
    console.log('无需回填');
    return;
  }

  const fetchScript = resolve(scriptDir, 'jiuyan-image-fetch.ts');
  let success = 0;
  let fail = 0;
  let skip = 0;

  for (let i = 0; i < dates.length; i++) {
    const date = dates[i];
    console.log(`[${i + 1}/${dates.length}] ${date}...`);

    if (dryRun) {
      console.log(`  (dry-run) 跳过`);
      continue;
    }

    try {
      const result = spawnSync('npx', ['--yes', 'tsx', fetchScript, date], {
        cwd: scriptDir,
        timeout: 200_000,
        encoding: 'utf8',
        stdio: ['pipe', 'pipe', 'pipe'],
        env: { ...process.env },
      });

      if (result.status !== 0) {
        throw new Error(result.stderr || `子进程退出码 ${result.status}`);
      }

      const data = JSON.parse(result.stdout.trim());
      const themes = data.themes || [];

      if (themes.length === 0) {
        console.log(`  无题材数据（可能非交易日），跳过`);
        skip++;
        continue;
      }

      // 写入 DB
      const record = {
        id: randomUUID(),
        pick_date: date,
        themes,
        raw_image_url: data.raw_image_url || null,
        source: 'jiuyan-image-backfill',
        created_at: Date.now(),
      };

      await sb.from('limitUpReasons').insert(record);
      console.log(`  写入成功: ${data.theme_count} 题材, ${data.stock_count} 股票`);
      success++;

      // 间隔 3 秒，避免限流
      if (i < dates.length - 1) {
        await new Promise(r => setTimeout(r, 3000));
      }
    } catch (err: unknown) {
      const msg = err instanceof Error ? err.message : String(err);
      console.log(`  失败: ${msg.slice(0, 150)}`);
      fail++;

      // 连续失败 3 次停止
      if (fail >= 3 && success === 0) {
        console.error('\n连续失败，停止回填。请检查 JIUYAN_SESSION 是否有效。');
        break;
      }

      // 失败后等 5 秒
      await new Promise(r => setTimeout(r, 5000));
    }
  }

  console.log(`\n回填完成: 成功 ${success}, 失败 ${fail}, 跳过 ${skip}`);
}

main().catch(err => {
  console.error('回填异常:', err);
  process.exit(1);
});
